UserProfile.js

```diff
@@ -0,0 +1,259 @@
+        // validate keyValue
+        if (keyValue === null || !Validate.isValidPrimitiveOrArray(keyValue)) {
+            L.w(`${functionName}, provided keyValue is not valid`);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * 
+     * Set custom key and value pair for the current user
+     * 
+     * @param {string} keyName - user property key
+     * @param {object} keyValue - user property value
+     * @returns {void}
+     */
+    setProperty = async function (keyName, keyValue) {
+        if(!this.isValidUserProfileCall(keyName, keyValue, "setProperty")) {
+            return;
+        }
+        let formattedKeyValue = keyValue.toString();
+        L.d(`setProperty, Setting user property: [${keyName}, ${formattedKeyValue}]`);
+        await this.#state.CountlyReactNative.userData_setProperty([keyName, formattedKeyValue]);
+    };
+
+    /**
+     * 
+     * Set predefined and/or custom key and value pairs for the current user
+     * 
+     * @param {object} userData - custom key value pairs
+     * @returns {void}
+     */
+    setProperties = async function (userData) {
+        if (!this.#state.isInitialized) {
+            L.w("setProperties, 'init' must be called before 'setProperties'");
+            return;
+        }
+        if (!userData) {
+            L.w("setProperties, User profile data should not be null or undefined");
+            return;
+        }
+        if (typeof userData !== "object") {
+            L.w(`setProperties, unsupported data type of user data '${typeof userData}'`);
+            return;
+        }
+        L.d(`setProperties, Setting properties: [${JSON.stringify(userData)}]`);
+        const predefinedKeys = {
+            name: "string",
+            username: "string",
+            email: "string",
+            organization: "string",
+            phone: "string",
+            picture: "string",
+            picturePath: "string",
+            gender: "string",
+            byear: "number",
+        };
+        const userProfile = {};
+        for (const key in userData) {
+            const value = userData[key];
+            const expectedType = predefinedKeys[key];
+            if (expectedType) {
+                if (typeof value === expectedType || (key === "byear" && typeof value === "number")) {
+                    userProfile[key] = key === "byear" ? value.toString() : value;
+                } else {
+                    L.w(`setProperties, skipping key '${key}' due to type mismatch (expected: ${expectedType}, got: ${typeof value})`);
+                }
+            } else {
+                if (Validate.isValidPrimitiveOrArray(value)) {
+                    userProfile[key] = value;
+                } else {
+                    L.w(`setProperties, skipping custom key '${key}' due to unsupported data type '${typeof value}'`);
+                }
+            }
+        }
+        await this.#state.CountlyReactNative.setProperties([userProfile]);
```

Comment:
what does [] do? is not userProfile is already a map kinda object

Reply:
java function takes ReadableArray as a parameter, it's required to pass in this way

Reply:
oh I see